### DISCUSSION TITLE
UVJ format / Open (M)SLA format

### DISCUSSION THREAD

**Discussion by Spiritdude** [sn4k3/UVtools#687] [Ideas]
I tried to find information for the UVJ format, beside the source code of UVtools. I found a brief description at https://github.com/ezrec/uv3dp/tree/master/uvj - it seems as if @ezrec defined the format?

What UVJ is missing is the ability to expose multiple images at the same Z, e.g. multi exposure support, which is helpful when printing calibration pieces in one go to find best print settings for a resin or environmental condition (like temperature).

## Gathering Thoughts
**I'm in the midst to gather my thoughts on a Open (M)SLA format**, something like .SL1 or .UVJ, but with multi-exposure support:
- each exposure image (instead of layer image) has its own settings incl. Z
- I have been pondering on incorporating a `main.gcode` which describes the entire print process using G-code:
  - Z motion, by using `G1 Z{lift_distance}` the firmware would be still in charge of some settings
  - UV light, e.g. using `M356` code alike Marlin's `M355`
  - custom G-code to load exposure image to LCD, like `M1200 I"img0001.png"`

## Goals
- Multi-exposure capability
- Open format (well documented, easy to implement on encoding/decoding stage)
- Slicing a model once, and have it print on various MSLA machines with different XY resolutions but similar build volume

## Pixels vs Vectors
For now I tend to think in terms of pixels per exposure image, but in order to support SLA as well, images could be also vector-based (exterior and interior polygons).

The advantage of vector based images would be to a model slice once, and have it print on different SLA or MSLA machines. 
- SLA with different laser point sizes would still calculate exact offsets to trace the walls and filling process (almost like FDM)
- MSLA with different XY resolutions would need to rasterize it to their native resolution

## Burden on Firmware/Controller
The simplest is the RLE format to pack pixels, like the CTB or PW* formats do, but each machine needs its own format - the burden on the controller and firmware is minimal, just unpacking pixels for the LCD to expose UV light.

The unpack of PNG exposure images is a bigger task, as it requires `libpng` to support the many formats (bw, gray, rgb(a) and packing methods), not sure about runtime RAM requirements involved. 

To rasterize polygons with holes is no easy task for a microcontroller, the complexity direct relates to required RAM as in data structures and/or stacksize involved.

Which leads to the requirement on the controller, a small Atmel running Marlin vs RPi/SBC which runs Linux, which easily can rasterize vector slices, and those many 32bit ARM controllers in between.

## Thoughts, Comments, Links

Constructive thoughts, comments and links to useful resources are appreciated.

**Comment by sn4k3** [comment]
> I found a brief description at https://github.com/ezrec/uv3dp/tree/master/uvj - it seems as if @ezrec defined the format?

Correct.

> What UVJ is missing is the ability to expose multiple images at the same Z, e.g. multi exposure support, which is helpful when printing calibration pieces in one go to find best print settings for a resin or environmental condition (like temperature).

It can, like other formats UVtools create a new layer in same position Z, UVJ defines `PositionZ` per layer which allow you to have unlimited layers in same position. Layer number is linked to imageN.png

A good and future proof format should be gcode + png images. Forget about weak microcontrollers, microprocessors with Linux and added functions and capabilities are the way to go, even if it adds a little more cost. Just like we see on FDM, most people add a Pi to have octoprint and other features, that is more expensive than provide a final board that is able to do that from start.
If you want to go microcontroller route keep it gcode and use RLE.

**Pixels vs Vectors:** There is one format that UVtools supports that use vectors: SVGX. However, I see no advantage on that. It will support other printers but who wants to print a file sliced with given settings for a given resin profile? That's a user error. Even on FDM I never dare to print other's .gcode even if is for the same printer brand and material, not because I don't trust the file but because I don't trust the settings which are never optimal, same goes otherwise. So, in the end people should always slice their own to guarantee the best settings and success and then having a vector that should be converted to image after all is not a bonus. Also, if you go vector forget about AA or get ready to have a mess in the vectors due the brightness change.

**Reply by Spiritdude** [reply]
@sn4k3 It skipped my attention the layer in UVJ can have its own Z - thanks for pointing this out.

So, UVJ fits then my requirements to a high degree: {mesh} > [slicing] => {UVJ with PNGs} => [firmware preproc] => {G-code with PNGs} => [firmware G-code with PNG processing extension]

**Resin Specific Settings**: like you, I also slice for each (FDM) printer as none is the same like the other, different nozzles, different filament; yet, for MSLA the resin dependent settings seem commonly in the firmware/device, no matter the settings in pwma (Mono 4K) or pmx2 files (Mono X2), the machines doesn't care, it overrides what the slicer set => the files are only describing which voxels to cure, but speed of motion and exposure times are kept & defined in the firmware/device only. 

How to other MSLA machine behave here?

## UVJ Multi-Exposure Capability
Back to UVJ, so essentially I can compose a multi-exposure print like this then:
```
   "layers": [
      { "Z": 0.05, "Exposure": 1 },       // image0000.png
      { "Z": 0.05, "Exposure": 0.5 },     // image0001.png
      { "Z": 0.05, "Exposure": 0.5 },     // etc
     ....
      { "Z": 0.1, "Exposure": 1 },
     ....
   ]
```

correct?

I would then convert in my controller the UVJ into some G-code according the firmware or file-settings (lift/retract, exposure times, etc) - the main task is to bring the exposure image to the LCD.

## Gcode Mockup
A brief mockup of G-code to operate a MSLA with Marlin or RRF:
```
G90                       ; absolute positioning
G28 Z                     ; home Z position
G0 Z6 F60                 ; move to Z=6 with 1mm/s

;LAYER:1
G90                       ; absolute positioning
G0 Z0.05 F60              ; move platform to Z=0.05 with 1mm/s
G1000 I"layer00000.png"   ; load image
M356 P1 S255              ; turn on UV light 100%
G4 S{bottom_exposure}     ; wait 40s
M356 P0                   ; turn off UV light
G91                       ; relative positioning
G0 Z{retract1} F{speed1}  ; move platform to Z=+1.5 with 1mm/s
G0 Z{retract2} F{speed2}  ; move platform to Z=+6.0 with 6mm/s
G0 Z-{retract2} F{speed2} ; move platform to Z=-6.0 with 6mm/s

;LAYER:2
G90                       ; absolute positioning
G0 Z0.1 F60               ; move to Z=0.1 with 1mm/s
G1000 I"layer00001.png"   ; load image
M356 P1 S255              ; turn on UV light 100%
G4 S{bottom_exposure}     ; wait 40s
M356 P0                   ; turn off UV light
...

;LAYER:6
G90                       ; absolute positioning
G0 Z3.05 F60              ; move to Z=3.05 with 1mm/s
G1000 I"layer00005.png"   ; load image
M356 P1 S255              ; turn on UV light 100%
G4 S{normal_exposure}     ; wait 2.5s
M356 P0                   ; turn off UV light
G91                       ; relative positioning
G0 Z{retract1} F{speed1}  ; move platform to Z=+1.5 with 1mm/s
G0 Z{retract2} F{speed2}  ; move platform to Z=+6.0 with 6mm/s
G0 Z-{retract2} F{speed2} ; move platform to Z=-6.0 with 6mm/s 

;LAYER:7
G90                       ; absolute positioning
G0 Z3.1 F60               ; move to Z=3.1 with 1mm/s
G1000 I"layer00006.png"   ; load image 
M356 P1 S255
G4 S{normal_exposure}
M356 P0
G91
G0 Z{retract1} F{speed1}
G0 Z{retract2} F{speed2}
G0 Z-{retract2} F{speed2}

;LAYER:8
G90
G0 Z3.15 F60
...
```
RRF with a dedicated plugin supports custom gcode, and call external program on SBC, to put layer/image to the UV LCD.

I like to transform at least one of my Anycubic Photon Mono 4K so I can control it better, [Turbo Resin](https://github.com/nviennot/turbo-resin) gives some direction. Yet I would prefer to replace the mainboard with RAMPS and run [PyCNC](https://github.com/Nikolay-Kha/PyCNC), but controlling the UV LCD might be the hard to do.

**Reply by sn4k3** [reply]
> yet, for MSLA the resin dependent settings seem commonly in the firmware/device, no matter the settings in pwma (Mono 4K) or pmx2 files (Mono X2), the machines doesn't care, it overrides what the slicer set => the files are only describing which voxels to cure, but speed of motion and exposure times are kept & defined in the firmware/device only.

> How to other MSLA machine behave here?

So far as I know almost all printers respect settings set on file inside a range, settings you set matters, you can try set 20s to normal layers and your anycubic machine will follow that. Global settings are used (set on slicer), advanced formats are able to follow per layer settings, including anycubic. If you use global settings, it allows you to modify parameters via LCD.
So far, the only printers that use firmware settings are the Halot series, which in new firmware ask you if you want to use firmware/lcd settings or file settings.
FDM is more tolerant to settings, in resin even a 0.2s off can ruin your print, as it happens to me when printing @ 0.1mm, 2.8s caused delamination and weak model, but 3s prints perfectly strong, so this kind of fine-tunning and use your own settings are most important for mSLA.

> Back to UVJ, so essentially I can compose a multi-exposure print like this then:
> ...
> correct?

Correct. You can generate a UVJ with UVtools and open as zip to inspect the structure.

> Gcode Mockup

That looks correct, you can use UVtools and convert to a gcode format and inspect the GCode syntax under it tab, there are already some defined commands. For example export to Chitubox -> ZIP
I use an Uniz IBEE which prints via gcode.

> but controlling the UV LCD might be the hard to do.

That is the most difficult part to do and why mSLA is still locked. LCD's are hard to address and there are no documentation. Mono LCDs are produced to a brand and documents are private.
The easiest way to drive them will be using an HDMI converter and send images via HDMI as a normal PC but is hard to find a cheap option as even Pi4 can't handle 8K or custom resolutions without proper hacking

**Reply by Spiritdude** [reply]
## Mixed Bag
My experiences with Mono 4K (Firmware 0.0.11 vs 2.0.2) and Lychee Slicer (4.x and now 5.1.8) has been very inconsistent, it fell back to default settings, even though I set it in the slicer - so I kept checking the settings on the printer directly; I manually changed UV intensity to 50% and when another printing .pwma it kept that setting, so the printer setting override the sliced file settings; after few such encounters, I concluded I can't trust any settings I make in the slicer, but only what I see when the print has started. I'm aware the Mono 4K being entry-level resin printer, so my expectations have been minimal too.

I tried UVtools.Cmd to convert sl1 into pwma but it produced a bad file, the file structure looked good, but the layer images were scrambled (confirmed via Photon Workshop beside my own previewer), as soon I made more tests I will post a github Issue (not sure if half-compiled and binary unzipped UVtools interfered with each other). Either way, I will continue to mature my own Python-based tool to handle SL1, PW* and UVJ sufficiently to convert the files myself.

## Workflow

So I'm not yet set with the entire workflow of mine.

- Now: Lychee Slicer 5.1.8 (Linux) -> pwma and pmx2 files
- Next: Prusa Slicer 2.x (Linux) to produce support -> .SL1 -> UVtools.Cmd -> .pwma or .pmx2
- Future: Prusa Slicer 2.x (Linux) to produce support -> .STL -> my own slicer -> .uvj -> UVtools.Cmd or my own tools -> .pwma or .pmx2

The reason for .uvj step in between, I like to keep track of resin & machine on command-line, see my [Prynt3r](https://xyzdims.com/3d-printing/prynt3r/) tool (very experimental stage, as it's a rewrite), right now it's for FDM only, but I like to extend it to SLA as well, eventually being able to do:
```
% prynt3r -p photon1 slice form.stl -o form.pwma @anycubic-white-resin
```

Whereas `photon1` is a label of the printer, which has a profile which defines the build-size and other specs. The `@` I use as a marker that it references a macro-file which lists multiple settings at once, e.g. in FDM I use `@revo-blue` `@pla+` which references a bunch of settings, kind of abbreviation.

As Mono 4K has no network, I need to transfer file manually - otherwise prynt3r logs all `print` with all settings incl. creating UUID of the print itself, right now with all the GUIs I have no protocol whatsoever, using `prynt3r` will allow me to keep track of this better (adding a switch to log also `slice` for example then). Just for sake of context, I attach NFC labels to some of my larger or complex FDM prints, those point to the prynt3r log, so I know exactly the settings I used, which printer, etc, and I like to do the same with SLA prints.

**Comment by Spiritdude** [comment]
# UVJ 1.0 Proposal - Vendor Neutral MSLA Format

Feel free to comment, suggest alternatives or additional keys etc.

As I realized, @sn4k3 / Tiago has extended silently the UVJ format as proposed by @ezrec / Jason by following new settings:
- `LiftHeight2`
- `LiftSpeed2`
- `RetractHeight2`
- `RetractSpeed2`

which implies the 2nd stage of build-plate movement, beside the peeling motion. I support this addition.

I propose following addition:
- new key called "Format", stating the format version like `"UVJ/1.0"`; the reason to use `"UVJ"` is to enforce a string, by having `"1.0"` alone might trick encoders/decoders to convert to float, and when for some reason we ever have `"1.10"` being interpreted as `1.1` we are running into troubles - so `"UVJ/1.0"` is clear enough
  `"Format": "UVJ/1.0"`
- any other keys not defined by UVJ Format in `config.json` are ignored, but preferably preserved at export; the reason for this is to allow vendors to add their own custom keys, e.g. I like to list all slicer settings in the .uvj file alike adding slicer settings at head or tail of .gcode files
  - pro: invites custom settings
  - contra: future version of UVJ may clash with custom settings, but as we enumerate the format, the encoder/decoder can work around it

Any .uvj which has not the `"Format"` key in its `config.json` is considered pre-1.0 and follows the @ezrec proposal as stated at https://github.com/ezrec/uv3dp/tree/master/uvj

**Comment by ezrec** [comment]
I highly recommend that the format version number follows the “semver” spec, specifically:

- Major number: New required fields are added, or semantics or spelling of an existing field changes.
- Minor number: New capabilities are added to the spec (ie, SVG supplemental source images for laser machines, per layer temp control, etc)
- Patch number: Optional fields added to existing sections.

See https://semver.org/ for the general guidelines.

The “Vendor” keys should be in a vendor block that can be added to any section, further wrapped by the vendor name, ie:

….
   Vendor: { Prusa: {
         StageTiltSpeed: 33,
         StageTiltAngle: 4
   } }
…

**Reply by Spiritdude** [reply]
@ezrec 
- major, minor I agree, about additional patch number I'm a bit skeptical to scare parties away from adoption, as it indicates for me more complexities of the format to parse or comply with
- the "Vendor" to be included in any section is an excellent idea, it's quite a powerful option to include then layer specific extension via "Vendor" section.

Here a sample including @sn4k3 extensions and my "Properties.Vendor.VoxGLSlicer" where I list the settings from my slicer:

```
{
   "Format": "UVJ/1.0",
   "Properties": {
      "Size": {
         "X": 3840,
         "Y": 2560,
         "Millimeters": {
            "X": 132.0,
            "Y": 80.0
         },
         "LayerHeight": 0.05,
         "Layers": 399
      },
      "Exposure": {
         "LightOnTime": 2.0,
         "LightOffTime": 0.5,
         "LiftHeight": 1.0,
         "LiftSpeed": 60.0,
         "RetractHeight": 1.0,
         "RetractSpeed": 60.0,
         "LiftHeight2": 5.0,
         "LiftSpeed2": 180.0,
         "RetractHeight2": 5.0,
         "RetractSpeed2": 180.0
      },
      "Bottom": {
         "Count": 5,
         "LightOnTime": 30.0,
         "LightOffTime": 0.5,
         "LiftHeight": 1.0,
         "LiftSpeed": 60.0,
         "RetractHeight": 1.0,
         "RetractSpeed": 60.0,
         "LiftHeight2": 5.0,
         "LiftSpeed2": 180.0,
         "RetractHeight2": 5.0,
         "RetractSpeed2": 180.0
      },
      "Vendor": {
         "VoxGLSlicer": {
            "verbose": 0,
            [ .... ]
            "layer_height": 0.05,
            "epsilon": 0.0001,
            "screen_width": 3840,
            "screen_height": 2560,
            "screen_modula": 0.005,
            "window": false,
            "window_width": 640,
            "format": "uvj",
            "output": "a.uvj",
            "machine_width": "132",
            "machine_depth": "80",
            "exposure_time": "2",
            "bottom_layers": "5",
            "bottom_exposure_time": "30",
            "screen_pixel": 0.035
         }
      }
   }
}
```

**Reply by sn4k3** [reply]
I agree with semver which most languages can handle easy, and they are strings. 1 or 1.2 or 1.2.1 are all valid and not forced to use all the dots. Also, instead consider version in an increment int like other formats do which is the simplest.
The `Vendor` dict is a good idea, I don't know how static languages serializers will handle with the missing methods losing all undeclared properties on serialize (UVtools case -- I may play latter with a List of dynamic and see if handles that) 
Edit: `Vendor` dictionary works well with C# object and dynamic, preserving properties and values

**Comment by Spiritdude** [comment]
A few questions of clarification for @ezrec & @sn4k3 :

## Two Stages of Motion
- `LiftHeight2`: lifting 2nd stage, `LiftHeight` is peeling off
- `RetractHeight2`: retracting 2nd stage, the `RetractHeight` retract to previous position

Now, what is the common practice here, as `LiftHeight` + `LiftHeight2` >= `RetractHeight` + `RetractHeight2`; when both sums are the same, the cured layer is push back to 0 distance to the FEP film, is this something wanted or even needed? 

If `RetractHeight` + `RetractHeight2` is larger, the cured piece will indent into the FEP film, correct?

## Lift vs Retract

What is the reason to differentiate between `LiftHeight` and `RetractHeight`? When is it not the same value?

## One Stage of Motion

@sn4k3: When `LiftHeight2` and `RetractHeight2` are not set, it's a single Z motion Lift and Retract, correct?

## Preview Tiny & Huge

@ezrec in your samples you indicated 
- `tiny.png`: 400x400 px (square)
- `huge.png`: 800x480 px (landscape)

Is this some guidance of size, or how shall we define tiny vs huge in pixels, and the aspect ratios?
Or keep sizes undefined and everyone can interpret it as they wish?

Also, the files are optional?

Suggested recommendation: "if previews are generated, make background transparent" or so.

## How to deal with additional files?

To adapt @ezrec "Vendor" for `config.json`, we also could allow `Vendor/Prusa/*` containing additional files like:
```
config.json
slice/
    000000.png
    000001.png
    ...
preview/
    tiny.png
    huge.png
Vendor/
   Prusa/
      scene.3mf
```

Thoughts?

**Reply by sn4k3** [reply]
> Now, what is the common practice here, as LiftHeight + LiftHeight2 >= RetractHeight + RetractHeight2; when both sums are the same, the cured layer is push back to 0 distance to the FEP film, is this something wanted or even needed?
>
> If RetractHeight + RetractHeight2 is larger, the cured piece will indent into the FEP film, correct?

`RetractHeight2` is the slow retract and set by user.
`RetractHeight` is never assigned by user and is calculated by the `TotalLift - RetractHeight2`, which can never be negative. In other words: `RetractHeight2 = Constrain(UserValue, 0, TotalLift)` and `RetractHeight = TotalLift - RetractHeight2`.
UVtools always sanitize values, if you put a larger value than TotalLift it will constrain. Printer firmwares does the same (I guess) just like the slicers.

Note that:

- First Lift is the slow lift
- Second Lift is the fast lift
- First Retract is the fast retract
- Second Retract is the slow retract


> When LiftHeight2 and RetractHeight2 are not set, it's a single Z motion Lift and Retract, correct?

Yes and no. Some printers have a second lift set on firmware, if you set 0mm it will use the firmware default. The problem with 0 values on firmwares is that they are never 0. For example, if you set 0mm lift's it will use firmware default, same goes for speed and everything else.

**Reply by Spiritdude** [reply]
(continuing clarification of terminologies & notions)

## Retract Heights

RetractHeight & RetractHeight2: 

I just checked the firmware of my Mono X2, there is 
- Lift Height 1 (slow)
- Lift Height 2 (fast)
- Lift Speed 1 (slow)
- Lift Speed 2 (fast)
- Retract Speed 1 (fast)
- Retract Speed 2 (slow)
 
the Retract Heights are implied to be the same as Lift Height 1 & 2, so they don't appear.

The Lychee Slicer and Photon Workshop does the same, there are no Retract Heights 1 & 2, as it is implied to be the same as Lift Height 1 & 2. 

So, if the format supports it, then RetractHeight 1 & 2 is something like fine-tuning descending the build-plate, and the firmware has to implement the constraints for RetractHeight2 arrive to the next layer Z pos as @sn4ke mentioned above.


## Light Off Time

@ezrec defines `LightOffTime` as "Amount of time for the light to be off before lift & retract, in seconds", this would mean right "before lift & retract" this means **after** light on and **before lift & retract**.

_(click on image to zoom in)_

![image](https://user-images.githubusercontent.com/2078766/232230973-31e28265-5d9b-4b73-9197-5871cd3b1a0d.png)


ChituBox Light Off Time: **"Light-off Delay = Building Plate Up and Down Time + Additional Delay Time"** as from https://helpcenter.phrozen3d.com/hc/en-us/articles/6371494663833-How-can-Light-off-Delay-improve-3D-prints-

The problem with this definition is that CB's "light off delay" is a minimum time, in the worst case the resin is still disturbed & in-motion after descend to Z pos, without any waiting time or "additional delay time".

![image](https://user-images.githubusercontent.com/2078766/232231141-cb528efb-4247-4aaa-a13f-5915ac0798d0.png)

When I checked ChituBox Interlayer G-code I saw this:
```
M6054 {image}; show image
G0 Z{rise_pos} F{rise_speed}
G0 Z{fall_pos} F{fall_speed}
G4 P{light_delay}; <====================
M106 S{light_pwm}; light on
G4 P{exposure_time}
M106 S0; light off
```
which makes it even more confusing, that the "Light-Off Delay" in the ChituBox slicer is not the same as `{light_delay}`.

![image](https://user-images.githubusercontent.com/2078766/232230988-8ccadee8-f364-4b4e-b2b1-3693d8c29a22.png)
 
Now, @ezrec 's defines the wait time after UV light (correct me if I misinterpret your definition), and looking at ChituBox and people tuning with "Light-Off Delay" their **printing experience**, we end up with at least 3 different times:
- before light on (after lift & retract)
- after light on (before lift & retract)
- light off time as minimum time for LCD not to get UV light, from the perspective of the LCD itself

From what I've got, these three times are mixed up and called "light-off time" or "light-off delay" confusing everyone, and this we need to resolve. As I was reading up and watching various videos on resolving the confusion, a 4th time was introduced, the wait time between lift & retract aka top Z retention time (e.g. https://www.youtube.com/watch?v=u0V961evYAc) which appear in some ChituBox G-code, which was fine-tuned.

So, to sum up, we have 4 times:
- before light on: `Wait(Time)BeforeLight` [s]
- after light on: `Wait(Time)AfterLight` [s]
- after lift before retract: `Wait(Time)AfterLift` [s]
- `LightOffTime` [s]: minimum time the LCD has no UV
  - `LightOffTimePosition`: `BeforeLight` (default), `AfterLight`, `AfterLift`; if the `LightOffTime` is not met by all the other timings, define where the additional compensations is made

I didn't draw the `LightOffTime` that is the +1 timing.

![image](https://user-images.githubusercontent.com/2078766/232233944-cbdea64a-b61b-44a7-8dcc-a03d1fa997f4.png)

Since ChituBox 1.9 they came up with the same, just naming the reference a bit differently (discovered this after I made all the graphics above): 

![Screenshot from 2023-04-15 17-08-08-cropped](https://user-images.githubusercontent.com/2078766/232233191-0d40e097-ef0f-4e74-9df1-ccb4aa547954.png)

Any additional development regarding timing?

**Comment by ezrec** [comment]
To be honest; my comments on those fields were best guesses at the time - I neglected to do any experiments to verify my assumptions.

**Comment by nviennot** [comment]
Hello everyone, author of https://github.com/nviennot/turbo-resin

File format
------------
I would much prefer that we don't use PNG. I've measured the ROM footprint, and it's ~20KB. Not a huge deal, but the RAM footprint is massive: 50KB due to the decompression algorithm. It's awful. It can be done, but there's not going to be much left for the rest of the firmware (e.g., things like wifi support).

As a point of comparison, the ROM footprint for the CTB format is 0.3KB, and PWMA 0.3KB. So that's a pretty massive difference.

Color depth
------------

We want to support at the very least 4-bit color depth. Note that it'd be nice to support people who want to use the 3D printer as a way to make circuit-boards.

Masks
-------

Things that are helpful when calibrating a new printer+resin is to do multiple exposures.
See the use of palette changes on the Mono 4K here: https://github.com/nviennot/reversing-mono4k/blob/main/writeup/part9/README.md#multiple-exposures (see last gif of that section).

Sending a 4K (or 8K) image to the LCD takes a *long* time, like 10secs at times, depending how well the image compresses.
This can be problematic for quickly applying mask for multiple exposures.

Maybe we want to have the ability to apply a given mask as an overlay, but that's something the FPGA driving the LCD panel should be able to understand. And it's hard to write a custom firmware for that.

Still, we should allow the file format to eventually specify masks.

G-code
--------

I think it's a good idea to use g-code. We could be specifying the acceleration/speed settings in a more standardized format. Maybe masks would be applied via a g-code command like "MXXX mask.png" or maybe "MXXY (x1,y1,x2,y2)" as a rectangle coordinate.

Also, I'd like the possibility to implement a force sensor in the printer (like https://blog.honzamrazek.cz/2022/01/prints-not-sticking-to-the-build-plate-layer-separation-rough-surface-on-a-resin-printer-resin-viscosity-the-common-denominator/) which is used by the firmware to decide on the correct build plate movement. The maximum "pressure" could be defined by g-code.

All the json-like parameters per-layer is wrong in my opinion. It doesn't compose well. We can't make experiments. Having different speeds is silly. What if I want 3? G-code is the answer.

**Reply by sn4k3** [reply]
**UVJ** is a format made to be "human friendly" to view, inspect, hand edit and then convert to target format. Was not intent to be used on printers. It can be used but not with low spec printers due PNG. Workflow: Convert a format to UVJ -> view/edit -> Convert back to original format -> Print -- Was the original motive of UVJ.

If you are looking for a format that handles everything in binary / RLE format look at [OSLA](https://github.com/sn4k3/UVtools/blob/master/documentation/osla.md), it was composed by me, supports a ton of image types and uses gcode. While it still records global and per layer data in its structure, it is pure informative, all parameters are on gcode and should be parsed from there. Have a look at it.

**Reply by Spiritdude** [reply]
We are mixing motives now, here my motive:
1) I thought of UVJ to be a good base for an Open MSLA File Format, as it seemed suitable (I prefer JSON over XML) as format(s) are well defined (PNG, JSON)
2) UVJ in its current format lacks a few features (that's we are discussing here)
3) I wasn't aware of OSLA format, which covers some of @nviennot input
4) I want to use UVJ on my printer direct, and if I need to replace the controller and use a SBC so be it - or OSLA if it covers multi-exposure feature, at the first glance it seems it does when each layer has its own Z value

**Reply by Spiritdude** [reply]
@sn4k3 what would be the feature differences between UVJ with the additions we discussed so far and your current OSLA 1.0 definition?

To ask specifically:
- vendor or custom specific settings could be added? would that be done via custom table? how would one separate vendor namespace?
- OSLA 1.0 supports optional full G-code control as I saw, which means variable layer heights, multi-exposures, multiple speeds etc. that way I can write G-code to fully control all parts: Z motion, UV image load/exposure, even UI LCD as long the firmware supports it, correct?

**Reply by sn4k3** [reply]
> vendor or custom specific settings could be added? would that be done via custom table? 

Yes, custom table is there to insert any values you want "without" limitation.

> how would one separate vendor namespace?

Binary files use only values, however you can use separators key-bytes, the custom table is defined by writer. Readers like UVtools keep the table but won't show it to user as the struct is unknown/custom.

> OSLA 1.0 supports optional full G-code control as I saw, which means variable layer heights, multi-exposures, multiple speeds etc. that way I can write G-code to fully control all parts: Z motion, UV image load/exposure, even UI LCD as long the firmware supports it, correct?

Correct. GCode is preferred if available, if printer unable to follow gcode or not present it will fallback to layer definition table which hold the backup values.

**Comment by Spiritdude** [comment]
## OSLA: Powerful, Requirement for Labeling Capabilities for Firmware/Device 

@nviennot the OSLA definition I glanced over, here my thoughts:
- do you see an [image format](https://github.com/sn4k3/UVtools/blob/master/documentation/osla.md#image-data-type) suitable for Turbo Resin?
- does some of image pointer and XY offsets & WH clipping feature cover some of your mask ideas?
- whoever writes `.osla` files has to have the option of the layer images encoding; and MSLA printer manufacturer have to state "OSLA:PNG+BMP" supported, and in case of Turbo Resin e.g. "OSLA:BMP" supported due the limited RAM available for decoding PNG
- we have G-code as the controlling core technology, so we are not limited: most of the base definitions and layer definitions with lift/retract speeds are "fallback" settings; in a way, "OSLA:PNG+BMP+GCODE" supported, or "OSLA:BMP+GCODE" supported, and perhaps some won't do G-code support, and then omit "GCODE" in the label.

The definition is powerful, but with it also makes it hard(er) to support, and labeling what exactly is supported in a firmware/device is required. And the slicers need to know which devices are capable which features of OSLA.

The idea to use a load cell to increase print speed as you wrote me by email, is there something the file format needs to support (UVJ or OSLA), or this is mainly a firmware feature [measure area exposed, and combine Z-motions with load cell measurement]?

**Reply by sn4k3** [reply]
OSLA is still a draft and I need to make some alterations (Also open to ideas), such as file is limited to 4GB due `uint` address to layer data and gcode, need to add pagenumber to solve that for low end boards. Also need to introduce some flag fields for future proof.

Is true that the versatility of image format is a "problem" for slicers and UVtools, maybe we should stick with one for previews like RGB24 and another for layers like RLE, instead of giving chooise? That will simplify and make it more easy and less messy to read/write.

> The idea to use a load cell to increase print speed as you wrote me by email, is there something the file format needs to support (UVJ or OSLA), or this is mainly a firmware feature [measure area exposed, and combine Z-motions with load cell measurement]?

Loadcell controls lifts, so in gcode all you need is to implement the loadcell M command to make the lifts by it measurements, eg:


| Traditional | Load cell | Comment |
|----------|-------------|-------------|
| M6054 "1.png"; | M6054 "1.png"; | Show image |
G0 Z7.1 F8; | M8001 L2.5 M10 S5 F10 Z0.1 T20000 | Lift |
G0 Z0.1 F10; | | Retract |
G4 P3000; | | Stabilization delay |
M106 S300; | M106 S300; | Led power |
G4 P20000; | G4 P20000; | Cure time |
M106 S0; | M106 S0; | Turn led off |
G4 P3000; |  G4 P3000;  | Wait to cooldown

M8001 is the command I just made to indicate lift is handled by loadcell
```
L = Minimum lift (Optional)
M = Maximum lift (Optional)
S = Slow feedrate (Optional)
F = Fast feedrate (Optional)
Z = Return position (Optional? and if so, writer must issue commands to return to Z position and all optional other delays)
T = Maximum time to wait for stabilize on Z position (Optional)
```

This is just an idea on how to implement the load cell using gcode.
For layer data all we need is to add a new field to indicate to use load cell or not

**Comment by yaqwsx** [comment]
@Spiritdude reached out to me via Twitter to express my opinion on the topic. TBH, I haven't read the discussion.

But I have been thinking about programming MSLA machines for a long time in past, and I came up with the following proposal (it is commented example program). I abandoned the idea of building a custom driver electronics recently, so I decided to share it publically:

```.sh
# There are line comments denoted by '#'. There is no other type of comment.

# There is a mandatory language type and version specification
language MSLALang version=1

# Then there can be a list of parameters that the user can set before starting
# the print. Parameters marked as mutable can be adjusted during runtime.
param baseExposure type=number default=2 min=0 max=10 mutable=true
param bottomExposure type=number default=20
param baseLayerCount type=number default=5

# We can also have constants that make it easy for humans to understand what's
# going on
const layerHeight type=number value=0.05

# A few words about types: There are two built-in types: number and string.
# Additional drivers can implement a new types. Usually that would be an image
# that we can project on the screen. However, you cannot reference these types
# from the MSLA language.

# A program in the MSLA language is just a set of function and procedures. One
# of the procedures must be called main. This is the procedure where the
# evaluation of the program starts.
#
# Most of the procedures are built-in or they are imported from other modules.

# Functions are expressions – they take arguments and return a value. They can
# invoke other functions, but they cannot invoke procedures. The functions
# cannot be recursive.
#
# Basically function are "just" named expressions.
#
# We distinguish two types: number and string. The strings can be formatted.
fun exposureTime(layerNumber: number) -> number:
    if layerNumber <= baseLayerCount
        then bottomExposure
        else baseExposure

# This is an example of a function that returns a layer slice name as a string
fun layerSliceName(layerNumber: number) -> string:
    "layerImage_${layerNumber}"

# Procedures are non-recursive subprograms that take arguments and perform
# actions on the machine.
#
# Each line is a single invocation of another procedure. The procedures take
# arguments via name=value. If the value should be computed, it has to be
# enclosed inside the argument evaluation section ${}
proc layerCycle(layerNumber: number):
    reportLayerStart layerNumber
    moveAbs Z=(${layerHeight * (layerNumber + 1)})
    # The retrieve function retrieves a slice image from the program archive.
    showPattern pattern=(retrieve (layerSliceName layerNumber))
    startExposure
    wait (exposureTime layerNumber)
    machinePeelCycle
    reportLayerEnd layerNumber

# Each program starts with a procedure main:
proc main():
    home
    moveAbs Z=0
    for i in 1..5000:
        layerCycle(layerNumber)
    moveSafeZ
    message "Print completed"

# You might be wondering - why such restrictions? Well, the language is designed
# to be decidable. We can decide if the program halts, we can simulate it and
# never run into an infinite loop. This is essential as we want to make precise
# time estimates and simulations of the printing process.

# To make the language useful, we should define a mandatory procedures for given
# types of machines. An example of such mandatory procedure is, e.g.,
# machinePeelCycle. It is a built-in procedure that peels a layer. Every machine
# has it. Dumb machines will just wait and lift; smart machines will use, e.g.,
# a force sensor to peel. Having the minimal set of required procedures allows
# you to write portable programs. However, if you know that you target a
# specific machine, you can use machine-specific procedures.

# The language is designed to be powerful, hackable and human readable, however,
# it comes at the cost that post-processing tools, such as UVTools, will have
# hard time interpreting the programs and modyfing them. However, the question
# is if using a post-processing is a good path in a long term anyways and if all
# the functionality UVTools offer shouldn't be migrated into a custom
# open-source slicer.
```
